### PR TITLE
Parse embedded variants in case of no slot definiton

### PR DESCRIPTION
**Pull request by gusman**
A fix for embedded variants issue.
I extract _parser function with name parse_variant_line to become public function and not an internal function of parse_custom_variant.
I reuse parse_variant_line to get slot information from lc timos_line.

### CODE REVIEW

File: sros/docker/launch.py [hunk 1]
```diff
@@ -827,10 +826,22 @@ def __init__(self, hostname, username, password, mode, variant_name, conn_mode):
             # LC VM Instantiation
             start_eth = 1
             lc_slot_tracker = []
+
             for lc in variant["lcs"]:
-                lc_slot = lc["slot"]
+                lc_slot = lc.get("slot", None)
```
Comment (hellt) [inline sros/docker/launch.py:831]: maybe lets have a fallback to `1`? 
```suggestion
                lc_slot = lc.get("slot", "1")
```

File: sros/docker/launch.py [hunk 2]
```diff
@@ -827,10 +826,22 @@ def __init__(self, hostname, username, password, mode, variant_name, conn_mode):
...
+                    lc = parse_variant_line(timos_line, lc)
+                    self.logger.info(f"New reconsructed LC: {lc}")
+
+                    # Retrieve slot information from new constructed LC
+                    lc_slot = lc.get("slot", None)
+
 
-                # Validation of lc_slot value
+                # Final validation of lc_slot value
+                # If lc_slot does not exist the skip instantiation
```
Comment (hellt) [inline sros/docker/launch.py:844]:
```suggestion
                # If lc_slot does not exist - skip instantiation
```

File: sros/docker/launch.py [hunk 3]
```diff
@@ -827,10 +826,22 @@ def __init__(self, hostname, username, password, mode, variant_name, conn_mode):
...
+
             for lc in variant["lcs"]:
-                lc_slot = lc["slot"]
+                lc_slot = lc.get("slot", None)
+
+                # Try to retrive from timos_line
+                if not lc_slot:
+                    timos_line = lc.get("timos_line")
+                    lc = parse_variant_line(timos_line, lc)
+                    self.logger.info(f"New reconsructed LC: {lc}")
```
Comment (hellt) [inline sros/docker/launch.py:837]: why don't we add the logic of parsing the variant line to line https://github.com/hellt/vrnetlab/blob/428ee4f70a887f8e3b438da39a9041275d7513aa/sros/docker/launch.py#L742

There we get the variant obj from the embedded variants, and after that we parse the variant lines for the retrieved object.
That way, we keep this code block of LC VM instantiation intact

Comment (gusman) [reply sros/docker/launch.py:837]: Do yo mean to update the all variant LC information before entering the LC VM instantiation code block?
I think that is doable. I will push again today for that changes.

File: sros/docker/launch.py [hunk 4]
```diff
@@ -322,24 +323,27 @@ def _parse(cfg, obj, skip_nics=False):
...
-        variant = _parse(cfg, obj=variant)
+        variant = parse_variant_line(cfg, obj=variant)
 
     return variant
 
 
 # sort line card definitions if they were provided in a non-consequetive order
 def sort_lc_lines_by_slot(lc_lines: list) -> list:
-    timos_tuples = [(v["slot"], v) for v in lc_lines]
+    max_slot = 99
```
Comment (gusman) [inline sros/docker/launch.py:345]: This part to ensure if no slot information then put it to the last sequence in tuples

File: sros/docker/launch.py [hunk 5]
```diff
@@ -741,6 +745,13 @@ def __init__(self, hostname, username, password, mode, variant_name, conn_mode):
 
         if variant_name.lower() in SROS_VARIANTS:
             variant = SROS_VARIANTS[variant_name.lower()]
+
```
Comment (gusman) [inline sros/docker/launch.py:748]: New implementation to re-parse timos_line in embedded variant

**Comment by hellt** [comment]
@gusman in 1bd5d68 I moved the `bof save` command to the if block where the bof is configured.
I noticed that otherwise the bof save was not applied, due to model driven mode was enabled

**Comment by hellt** [comment]
@gusman if you agree with that change, I am good with merging this PR
thanks for looking at that issue

File: sros/docker/launch.py [hunk 6]
```diff
@@ -550,16 +554,18 @@ def bootstrap_config(self):
...
                 "/configure system management-interface configuration-mode {mode}".format(
                     mode=self.mode
                 )
             )
 
-        # save bof config on disk
-        self.wait_write("/bof save")
-        self.wait_write("/logout")
+            # logout at the end of execution 
+            self.wait_write("/logout")
```
Comment (gusman) [inline sros/docker/launch.py:568]: I move the /logout command to at the end of boostraping config exexuction, to ensure all comand in boostraping config are executed before logout from terminal.

Comment (hellt) [reply sros/docker/launch.py:568]: thanks, good catch.

**Pull request merged**